### PR TITLE
Exception during iterator yield caused by multiple issues in LookupTask

### PR DESCRIPTION
**Pull request by billoley**

### CODE REVIEW

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (EdColeman) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:89]: Is this an error? If so, could the message also include what the error / issue is?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (EdColeman) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:94]: Question: Does this really walk the iterator - or is it always returning the first entry pointed at by the iterator?

**Review by EdColeman — COMMENTED** [review]
The code looks reasonable - but it is a section of the code that I don't have a full understanding right now to say that it is correct - I assume so.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (billoley) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java]: I decided that adding Pair<KeyExtent, List<Range>> of an unfinished Range back to the front of the LinkedList ensures that it gets handled next.  That means that if the LookupTask ends due to time or result size then a record of the last partial scan will get returned.  If the unfinished ranges get fully scanned, then the KeyExtent will be recorded as a full scan and the record of a partial scan will be removed.  This seems to make the most sense.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (billoley) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java]: Was just trying to get the first element, not iterating.  In any case, I modified the code to use a LinkedList.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (ivakegg) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:139]: How were we not always getting concurrent modification exceptions here?  Weird

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (ivakegg) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:84]: We need to make sure the session.queries is reset with what is left in the queryQueue after the loop.  Otherwise whenever the scan session is restarted (e.g. bytesAdded > maxResultsSize) then we need to restart where we left off.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (billoley) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:84]: Committed a change to remove and add to session.queries whenever a change is made to queryQueue.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (billoley) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:139]: CME occurred the next time iter.next() was called, so the while condition would have to allow one more execution.  My theories are:  1) low number of yields 2) only one extent 3) yield happens on last extent 4) timeout in while coincides with yield

File: server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java
Comment (ivakegg) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:119]: I wonder if a comment here on why we are doing it this way since the reason is non-obious.

**Review by ivakegg — APPROVED** [review]

**Comment by ctubbsii** [comment]
Can we get a bit more information about what the problem is that this is solving? There seems to be a lot of discussion about the fix, but I don't understand the nature of the problem. What was the exception that was seen, and under what general circumstances? What is the high level summary of the problem that this is fixing?

**Comment by billoley** [comment]
@ctubbsii -- The original issue and those that I found while fixing that issue are described in the issue -- https://github.com/apache/accumulo/issues/2789 -- The exception in Tablet/TabletBase ("Underlying iterator yielded to a position that does not follow the last key returned: " ) was the first one encountered.   Let me know if I should add more detail in the issue.

**Comment by ctubbsii** [comment]
@billoley Thanks. I didn't see the reference to the original issue in the PR description or subject line, so didn't make the connection. I think the original issue sufficiently describes the situation observed in detail. However, it doesn't quite explain the nature of these changes, which we'll need summarized for the squashed git commit when we merge this. If you have a brief summary of the changes being done om the PR, that will allow the reader to connect back to the more detailed description contained in the referenced issue, that would be helpful.

Also, are you anticipating creating a backport PR for 1.10 after this is merged to 2.1?

**Comment by billoley** [comment]
Rebased on main and force-pushed to the MR branch to resolve a conflict (that required a new method in ScanSession)

**Comment by dlmarion** [comment]
Full IT build passed successfully

**Comment by ctubbsii** [comment]
@ivakegg Next time, please squash the PR commits to merge with a clean history. That's been our routine practice for a long time, with relatively few exceptions. The option in GitHub is called "Squash and Merge". If it's not the default for you in the big green merge button, then you can select it by choosing the arrow next to the merge button.

**Comment by ctubbsii** [comment]
Repeating my earlier question:

> Also, are you anticipating creating a backport PR for 1.10 after this is merged to 2.1?

**Comment by ctubbsii** [comment]
> @billoley Thanks. I didn't see the reference to the original issue in the PR description or subject line, so didn't make the connection. I think the original issue sufficiently describes the situation observed in detail. However, it doesn't quite explain the nature of these changes, which we'll need summarized for the squashed git commit when we merge this. If you have a brief summary of the changes being done om the PR, that will allow the reader to connect back to the more detailed description contained in the referenced issue, that would be helpful.

@ivakegg @billoley My comment above seems to have been completely ignored, when I suggested that the commit message need to be improved when this is merged as a squashed commit. None of the 7 commits added a concise description of the changes, and the last commit message did not even follow git log conventions, having a subject line that is long enough to wrap my wide screen. Yet it was merged, as-is, anyway. Please consider reviewing the suggestions in https://cbea.ms/git-commit/ for writing good commit messages.

**Pull request merged**
